azurekvp/src/tracing.rs

```diff
@@ -0,0 +1,154 @@
+pub fn handle_span<S>(
+    span: &tracing_subscriber::registry::SpanRef<'_, S>,
+    file_path: &Path,
+    status: &str,
+    event_level: &str,
+    end_time: SystemTime,
+) where
+    S: tracing::Subscriber
+        + for<'lookup> tracing_subscriber::registry::LookupSpan<'lookup>,
+{
+    let span_context = span.metadata();
+    let span_id = Uuid::new_v4();
+
+    if let Some(start_instant) = span.extensions().get::<MyInstant>() {
+        let elapsed = start_instant.elapsed();
+
+        let start_time = end_time
+            .checked_sub(elapsed)
+            .expect("SystemTime subtraction failed");
+
+        let start_time_dt = DateTime::<Utc>::from(
+            UNIX_EPOCH
+                + std::time::Duration::from_millis(
+                    start_time.duration_since(UNIX_EPOCH).unwrap().as_millis()
+                        as u64,
+                ),
+        )
+        .format("%Y-%m-%dT%H:%M:%S%.3fZ")
+        .to_string();
+
+        let end_time_dt = DateTime::<Utc>::from(
+            UNIX_EPOCH
+                + std::time::Duration::from_millis(
+                    end_time.duration_since(UNIX_EPOCH).unwrap().as_millis()
+                        as u64,
+                ),
+        )
+        .format("%Y-%m-%dT%H:%M:%S%.3fZ")
+        .to_string();
+
+        let event_value = format!(
+            "Start: {} | End: {} | Status: {}",
+            start_time_dt, end_time_dt, status
+        );
+
+        handle_kvp_operation(
+            file_path,
+            event_level,
+            span_context.name(),
+            &span_id.to_string(),
+            &event_value,
+        );
+    }
+}
+
+/// Handles event data by encoding the message and metadata as key-value pairs (KVP)
+/// and writing the encoded data to the specified log file.
+///
+/// # Arguments
+/// * `event_message` - A string message associated with the event.
+/// * `span` - The span associated with the event, used to retrieve span context and metadata.
+/// * `file_path` - The path to the log file where the encoded KVP data should be written.
+/// * `event_instant` - A `MyInstant` representing when the event occurred.
+pub fn handle_event<S>(
+    event_message: &str,
+    span: &tracing_subscriber::registry::SpanRef<'_, S>,
+    file_path: &Path,
+    event_instant: MyInstant,
+) where
+    S: tracing::Subscriber
+        + for<'lookup> tracing_subscriber::registry::LookupSpan<'lookup>,
+{
+    let span_context = span.metadata();
+    let span_id: Uuid = Uuid::new_v4();
+
+    let event_time = SystemTime::now()
+        .checked_sub(event_instant.elapsed())
+        .expect("SystemTime subtraction failed");
+
+    let event_time_dt = DateTime::<Utc>::from(
+        UNIX_EPOCH
+            + std::time::Duration::from_millis(
+                event_time.duration_since(UNIX_EPOCH).unwrap().as_millis()
+                    as u64,
+            ),
+    )
+    .format("%Y-%m-%dT%H:%M:%S%.3fZ")
+    .to_string();
```

Comment:
It looks like this function is used exclusively in the `on_event` handler and that `event_instant` is created immediately before calling this function. Unless I'm missing something, to calculate the event time what we want is `SystemTime::now`. We also want to be careful and avoid panicking unless absolutely necessary. Since we have the span duration, we can fall back to using that if calculating the duration since the epoch fails, or failing that use the default value for Duration:

```rust
    // Calculate the current time, falling back to the elapsed time in the unreasonable event that the
    // system time is before the epoch.
    let event_time = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap_or_else(|_| {
            span.extensions()
                .get::<MyInstant>()
                .map(|instant| instant.elapsed())
                .unwrap_or_default()
        });
    // Note that we don't need the `to_string()` since we immediately pass it on to `format!()`
    let event_time_dt = DateTime::<Utc>::from(UNIX_EPOCH + event_time)
        .format("%Y-%m-%dT%H:%M:%S%.3fZ");
```